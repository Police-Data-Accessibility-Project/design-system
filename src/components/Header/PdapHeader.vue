<template>
	<header ref="el" class="pdap-header">
		<a v-if="!navLogoLinkIsPath" :href="logoImageAnchorPath" class="logo"
			><img
				:src="logoImageSrc"
				loading="lazy"
				width="250"
				alt="Police Data Accessibility Project Logo"
		/></a>
		<router-link v-else :to="logoImageAnchorPath" class="logo"
			><img
				:src="logoImageSrc"
				loading="lazy"
				width="250"
				alt="Police Data Accessibility Project Logo"
		/></router-link>
		<PdapNav />
	</header>
</template>

<script setup lang="ts">
import { onBeforeUnmount, onMounted, ref } from 'vue';
import { RouterLink } from 'vue-router';
import PdapNav from '../Nav/PdapNav.vue';
import { PdapHeaderProps } from './types';
import lockup from '../../assets/lockup.svg';

// Props
const props = withDefaults(defineProps<PdapHeaderProps>(), {
	logoImageSrc: lockup,
	logoImageAnchorPath: '/',
});

// Vars
const el = ref<HTMLElement | null>(null);
const navLogoLinkIsPath = props.logoImageAnchorPath.startsWith('/');

// Lifecycle methods
onMounted(() => {
	getHeightAndSetToCSSVar();
	window.addEventListener('resize', getHeightAndSetToCSSVar);
});

onBeforeUnmount(() => {
	window.removeEventListener('resize', getHeightAndSetToCSSVar);
});

// Utils
function getHeightAndSetToCSSVar() {
	if (el.value) {
		(document.querySelector(':root') as HTMLElement)?.style.setProperty(
			'--header-height',
			`${el.value.clientHeight}px`
		);
	}
}
</script>

<script lang="ts">
/**
 * # `Header`
 *
 * ## Props
 * @prop {string} logoImageSrc Src for the PDAP logo image to be displayed
 * @prop {string } logoImageAnchorPath Path for the link that wraps the PDAP logo image
 *
 * ## Notes
 * The `Header` component does not require any props to be passed. But keep in mind that it is responsible for rendering the `Nav` component. Consuming applications will need to `provide` an array of nav links -- **there are no defaults for this**, you must `provide` these links either 1\. in a layout component (see example below), at the route level, or at the app level. This allows for flexibility in which links are rendered on which routes
 *
 */
export default {
	name: 'PdapHeader',
};
</script>

<style>
.pdap-header {
	@apply dark:bg-neutral-300 relative bg-neutral-50 flex justify-between p-4 md:p-6 w-full z-50;
}

.logo {
	@apply decoration-0 relative max-w-[80%] px-4;
	@apply xs:max-w-none;
}

.logo img {
	@apply dark:invert;
}
</style>
