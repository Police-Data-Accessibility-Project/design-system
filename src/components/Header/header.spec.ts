// import * as VueRouter from 'vue-router';

// Component
import PdapHeader from './PdapHeader.vue';

// Utils
import { mount } from '@vue/test-utils';
import { describe, expect, test, vi } from 'vitest';

// Mocks
const base = {
	global: {
		provide: {
			navLinks: [
				{ path: '/', text: 'Home' },
				{ path: '/a', text: 'Home A' },
				{ path: '/b', text: 'Home B' },
				{ path: '', text: 'Link', href: 'https://www.google.com' },
			],
		},
		config: {
			warnHandler(msg, instance, trace) {
				if (
					msg.includes('Invalid prop: type check failed for prop') ||
					msg.includes('missing template')
				) {
					// ignore warning
					return;
				}
				console.warn(msg, instance, trace);
			},
		},
	},
};

vi.mock('vue-router');

// Test
describe('Header component', () => {
	// Render
	test('Renders a header', () => {
		const wrapper = mount(PdapHeader, base);

		expect(wrapper.find('.pdap-header').exists()).toBe(true);
		expect(wrapper.find('.pdap-nav').exists()).toBe(true);
		expect(wrapper.classes()).toContain('pdap-header');
		expect(wrapper.html()).toMatchSnapshot();
	});

	// Props
	// Props - logo src
	test('Renders header with default logo src', () => {
		const wrapper = mount(PdapHeader, base);
		expect(wrapper.props().logoImageSrc).toBe(
			'node_modules/pdap-design-system/dist/images/lockup.svg'
		);
	});

	test('Renders header with custom logo src', () => {
		const wrapper = mount(PdapHeader, {
			...base,
			props: { logoImageSrc: 'test' },
		});
		expect(wrapper.props().logoImageSrc).toBe('test');
	});

	// Props - anchor path
	test('Renders header with default logo anchor path', () => {
		const wrapper = mount(PdapHeader, base);
		expect(wrapper.props().logoImageAnchorPath).toBe('/');
	});

	test('Renders header with custom logo anchor path', () => {
		const wrapper = mount(PdapHeader, {
			...base,
			props: { logoImageAnchorPath: '/test' },
		});
		expect(wrapper.props().logoImageAnchorPath).toBe('/test');
	});

	test('Renders header with custom logo anchor href', () => {
		const wrapper = mount(PdapHeader, {
			...base,
			props: { logoImageAnchorPath: 'www.test.com' },
		});
		expect(wrapper.props().logoImageAnchorPath).toBe('www.test.com');
	});
});
