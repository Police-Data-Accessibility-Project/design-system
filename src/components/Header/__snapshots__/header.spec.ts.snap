// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > Renders a header 1`] = `
<header class="pdap-header">
  <!---->
  <div aria-controls="nav" aria-expanded="false" aria-label="open navigation menu" class="pdap-nav-open-button" role="button" tabindex="0">
    <transition-stub appear="false" css="true" persisted="true">
      <svg aria-hidden="true" class="svg-inline--fa fa-bars" data-icon="bars" data-prefix="fas" focusable="false" role="img" viewBox="0 0 448 512" xmlns="http://www.w3.org/2000/svg">
        <path class="" d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z" fill="currentColor" />
      </svg>
    </transition-stub>
  </div>
  <transition-stub appear="false" css="true" persisted="true">
    <nav aria-hidden="true" class="pdap-nav" id="nav" style="display: none;">
      <li class="pdap-nav-link-container">
        <!--v-if-->
        <!---->
      </li>
      <li class="pdap-nav-link-container">
        <!--v-if-->
        <!---->
      </li>
      <li class="pdap-nav-link-container">
        <!--v-if-->
        <!---->
      </li>
      <li class="pdap-nav-link-container">
        <a class="pdap-nav-link" href="https://www.google.com" referrerpolicy="no-referrer" target="_blank">Link</a>
        <!--v-if-->
      </li>
    </nav>
  </transition-stub>
</header>
`;
